File: Coin_gogogo/app/src/main/java/com/example/coin_gogogo/utility/CommentInfo_DiffUtil.java
package com.example.coin_gogogo.utility;

import android.util.Log;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;

import com.example.coin_gogogo.info.CommentInfo;
import com.example.coin_gogogo.info.PostInfo;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

import static com.example.coin_gogogo.utility.Named.HOUR;
import static com.example.coin_gogogo.utility.Named.MIN;
import static com.example.coin_gogogo.utility.Named.SEC;

public class CommentInfo_DiffUtil extends DiffUtil.Callback {

    private final ArrayList<CommentInfo> OldComments;
    private final ArrayList<CommentInfo> NewComments;
    private final PostInfo OldPost;
    private final PostInfo NewPost;

    public CommentInfo_DiffUtil(PostInfo OldPost, PostInfo NewPost) {
        this.OldPost = OldPost;
        this.NewPost = NewPost;
        this.OldComments = this.OldPost.getComments();
        this.NewComments = this.NewPost.getComments();
        Log.d("같z나zxc","old: " + OldPost.hashCode() + " new: "+ NewPost.hashCode());
    }

    @Override
    public int getOldListSize() {
        return OldComments.size();
    }

    @Override
    public int getNewListSize() {
        return NewComments.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        final CommentInfo oldpost = OldComments.get(oldItemPosition);
        final CommentInfo newpost = NewComments.get(newItemPosition);

        return oldpost.getKey().equals(newpost.getKey());
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) { //item이 같아도 수정된다면 내용이 다르다는 것을 인식시켜줘야 내용이 바뀜
        final CommentInfo oldpost = OldComments.get(oldItemPosition);
        final CommentInfo newpost = NewComments.get(newItemPosition);

        return oldpost.getContents().equals(newpost.getContents()) && oldpost.getPublisher().equals(newpost.getPublisher())
                && oldpost.getGood() == newpost.getGood();
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        //Implement method if you're going to use ItemAnimator
        return super.getChangePayload(oldItemPosition, newItemPosition);
    }

    public static String formatTimeString(Date postdate, Date nowDate){

        long ctime = nowDate.getTime();
        long regTime = postdate.getTime();

        long diffTime = (ctime - regTime) / 1000;
        String msg = null;

        if (diffTime < SEC) {
            msg = "방금 전";
        } else if ((diffTime /= SEC) < MIN) {
            msg = diffTime + "분 전";
        } else if ((diffTime /= MIN) < HOUR) {
            msg = new SimpleDateFormat("HH:mm").format(postdate);
//        } else if ((diffTime /= TIME_MAXIMUM.HOUR) < TIME_MAXIMUM.DAY) {
//            msg = (diffTime) + "일 전";
//        } else if ((diffTime /= TIME_MAXIMUM.DAY) < TIME_MAXIMUM.MONTH) {
//            msg = (diffTime) + "달 전";
        } else {
            msg = new SimpleDateFormat("MM월dd일").format(postdate);
        }
        return msg;
    }
}

File: Coin_kotlin/app/src/main/java/com/example/coin_kotlin/utility/CommentInfo_DiffUtil.kt
package com.example.coin_kotlin.utility

import android.util.Log
import androidx.recyclerview.widget.DiffUtil
import com.example.coin_kotlin.info.CommentInfo
import com.example.coin_kotlin.info.PostInfo
import com.example.coin_kotlin.utility.Named.HOUR
import com.example.coin_kotlin.utility.Named.MIN
import com.example.coin_kotlin.utility.Named.SEC
import java.text.SimpleDateFormat
import java.util.*


class CommentInfo_DiffUtil(
    val OldPost: PostInfo,
    val NewPost: PostInfo
) : DiffUtil.Callback() {

    private val OldComments = OldPost.comments
    private val NewComments = NewPost.comments

    override fun getOldListSize(): Int {
        return OldComments.size
    }

    override fun getNewListSize(): Int {
        return NewComments.size
    }

    override fun areItemsTheSame(
        oldItemPosition: Int,
        newItemPosition: Int
    ): Boolean {
        val oldpost: CommentInfo = OldComments[oldItemPosition]
        val newpost: CommentInfo = NewComments[newItemPosition]
        return oldpost.key.equals(newpost.key)
    }

    override fun areContentsTheSame(
        oldItemPosition: Int,
        newItemPosition: Int
    ): Boolean { //item이 같아도 수정된다면 내용이 다르다는 것을 인식시켜줘야 내용이 바뀜
        val oldpost: CommentInfo = OldComments[oldItemPosition]
        val newpost: CommentInfo = NewComments[newItemPosition]

        return (oldpost.contents.equals(newpost.contents) && oldpost.publisher
            .equals(newpost.publisher)
                && oldpost.good === newpost.good)
    }

    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
        //Implement method if you're going to use ItemAnimator
        return super.getChangePayload(oldItemPosition, newItemPosition)
    }

    companion object {
        fun formatTimeString(postdate: Date, nowDate: Date): String? {
            val ctime = nowDate.time
            val regTime = postdate.time
            var diffTime = (ctime - regTime) / 1000
            var msg: String? = null
            if (diffTime < SEC) {
                msg = "방금 전"
            } else if (SEC.let({ diffTime /= it; diffTime }) < MIN) {
                msg = diffTime.toString() + "분 전"
            } else if (MIN.let({ diffTime /= it; diffTime }) < HOUR) {
                msg = SimpleDateFormat("HH:mm").format(postdate)
                //        } else if ((diffTime /= TIME_MAXIMUM.HOUR) < TIME_MAXIMUM.DAY) {
//            msg = (diffTime) + "일 전";
//        } else if ((diffTime /= TIME_MAXIMUM.DAY) < TIME_MAXIMUM.MONTH) {
//            msg = (diffTime) + "달 전";
            } else {
                msg = SimpleDateFormat("MM월dd일").format(postdate)
            }
            return msg
        }
    }

}